src/diffusers/pipelines/pag/pag_utils.py

```diff
@@ -221,239 +143,82 @@ def _prepare_perturbed_attention_guidance(self, cond, uncond, do_classifier_free
-        return noise_pred
-
-    # Copied from diffusers.pipelines.pag.pag_utils.PAGMixin._prepare_perturbed_attention_guidance
-    def _prepare_perturbed_attention_guidance(self, cond, uncond, do_classifier_free_guidance):
-        """
-        Prepares the perturbed attention guidance for the PAG model.
-
-        Args:
-            cond (torch.Tensor): The conditional input tensor.
-            uncond (torch.Tensor): The unconditional input tensor.
-            do_classifier_free_guidance (bool): Flag indicating whether to perform classifier-free guidance.
-
-        Returns:
-            torch.Tensor: The prepared perturbed attention guidance tensor.
-        """
-
-        cond = torch.cat([cond] * 2, dim=0)
-
-        if do_classifier_free_guidance:
-            cond = torch.cat([uncond, cond], dim=0)
-        return cond
+        self._pag_attn_processors = pag_attn_processors
 
     @property
-    # Copied from diffusers.pipelines.pag.pag_utils.PAGMixin.pag_scale
     def pag_scale(self):
-        """
-        Get the scale factor for the perturbed attention guidance.
-        """
+        r"""Get the scale factor for the perturbed attention guidance."""
         return self._pag_scale
 
     @property
-    # Copied from diffusers.pipelines.pag.pag_utils.PAGMixin.pag_adaptive_scale
     def pag_adaptive_scale(self):
-        """
-        Get the adaptive scale factor for the perturbed attention guidance.
-        """
+        r"""Get the adaptive scale factor for the perturbed attention guidance."""
         return self._pag_adaptive_scale
 
     @property
-    # Copied from diffusers.pipelines.pag.pag_utils.PAGMixin.do_pag_adaptive_scaling
     def do_pag_adaptive_scaling(self):
-        """
-        Check if the adaptive scaling is enabled for the perturbed attention guidance.
-        """
+        r"""Check if the adaptive scaling is enabled for the perturbed attention guidance."""
         return self._pag_adaptive_scale > 0 and self._pag_scale > 0 and len(self.pag_applied_layers) > 0
 
     @property
-    # Copied from diffusers.pipelines.pag.pag_utils.PAGMixin.do_perturbed_attention_guidance
     def do_perturbed_attention_guidance(self):
-        """
-        Check if the perturbed attention guidance is enabled.
-        """
+        r"""Check if the perturbed attention guidance is enabled."""
         return self._pag_scale > 0 and len(self.pag_applied_layers) > 0
 
     @property
-    # Copied from diffusers.pipelines.pag.pag_utils.PAGMixin.pag_attn_processors with unet->transformer
     def pag_attn_processors(self):
         r"""
         Returns:
             `dict` of PAG attention processors: A dictionary contains all PAG attention processors used in the model
             with the key as the name of the layer.
         """
 
+        if self._pag_attn_processors is None:
+            return {}
+
+        valid_attn_processors = tuple(x.__class__ for x in self._pag_attn_processors)
```

Comment:
Curious on the choice of `tuple` as a data structure here. Why not a set in case performance is concerned.

Reply:
Ah yes, apologies for the oversight. Should be set